### PR TITLE
Add ffmpeg default thread options

### PR DESCRIPTION
**Pull request by kyamagu**
This PR adds threading support in ffmpeg ops.

- Set `codec_context_->thread_count =1` and `codec_context_->thread_type = FF_THREAD_SLICE` of [`AVCodecContext`](https://ffmpeg.org/doxygen/trunk/structAVCodecContext.html).

### CODE REVIEW

**Comment by yongtang** [comment]
@kyamagu I am wondering if we can set thread_type=2 internally so that user does not need to find out what will be the `ideal` value? For the majority of the users, they may have a hard time to find out what the thread_type means.

Also, from the performance result, it looks like `thread_count=1` will maximize the performance? Is there any reason behind it? If `thread_count=1` can maximize the performance in majority of the cases maybe we want to set this value as fixed as well?

**Comment by kyamagu** [comment]
@yongtang Sure, setting `thread_type=2` to be default makes sense.

As for the performance result, I believe there is a balance between threading overhead vs. performance gain, and likely it depends on the input video file (e.g., encoding or resolution). I will try checking more benchmark results.

**Comment by kyamagu** [comment]
I've generated test videos of 4 resolutions using the following:

```sh
ffmpeg -f lavfi -i testsrc=d=5:s=160x120:r=24,format=yuv420p -f lavfi -i sine=f=440:b=4 -shortest -y tests/test_video/test_160x120.mp4  
```

This time I check `thread_type IN {0, 1, 2}` and `thread_count IN {0, 1, 4}` combinations. Here is the benchmark results:

```
------------------------------------------------------------- benchmark 'ffmpeg_decode_video_benchmark_test_1280x960.mp4': 7 tests -------------------------------------------------------------
Name (time in ms)                Min                 Max                Mean             StdDev              Median                IQR            Outliers     OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
[test_1280x960.mp4-2-1]     522.0451 (1.0)      542.2541 (1.0)      530.0087 (1.0)       7.7530 (2.35)     528.1888 (1.0)       9.8806 (2.16)          2;0  1.8868 (1.0)           5           1
[test_1280x960.mp4-0-1]     552.5350 (1.06)     563.4701 (1.04)     556.2549 (1.05)      4.7598 (1.44)     553.5833 (1.05)      7.1390 (1.56)          1;0  1.7977 (0.95)          5           1
[test_1280x960.mp4-0-4]     611.7710 (1.17)     622.3560 (1.15)     619.3092 (1.17)      4.3658 (1.32)     620.5746 (1.17)      4.5819 (1.0)           1;0  1.6147 (0.86)          5           1
[test_1280x960.mp4-0-0]     614.9184 (1.18)     648.0066 (1.20)     628.5953 (1.19)     13.4946 (4.08)     627.7929 (1.19)     21.3383 (4.66)          2;0  1.5908 (0.84)          5           1
[test_1280x960.mp4-1-1]     624.4024 (1.20)     632.8130 (1.17)     627.7913 (1.18)      3.3050 (1.0)      626.6991 (1.19)      4.6240 (1.01)          2;0  1.5929 (0.84)          5           1
[test_1280x960.mp4-1-4]     626.8782 (1.20)     659.4065 (1.22)     634.5118 (1.20)     14.0316 (4.25)     627.4910 (1.19)     11.0943 (2.42)          1;1  1.5760 (0.84)          5           1
[test_1280x960.mp4-2-4]     633.3286 (1.21)     679.5432 (1.25)     654.6368 (1.24)     18.4623 (5.59)     654.6114 (1.24)     29.3866 (6.41)          2;0  1.5276 (0.81)          5           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------- benchmark 'ffmpeg_decode_video_benchmark_test_320x240.mp4': 7 tests -----------------------------------------------------------
Name (time in ms)              Min                Max               Mean            StdDev             Median                IQR            Outliers      OPS            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
[test_320x240.mp4-2-1]     39.8975 (1.0)      43.7382 (1.0)      42.2152 (1.0)      1.1760 (1.0)      42.7166 (1.0)       1.9696 (2.21)          8;0  23.6882 (1.0)          21           1
[test_320x240.mp4-0-4]     42.1934 (1.06)     64.0940 (1.47)     52.4384 (1.24)     8.4623 (7.20)     50.2504 (1.18)     16.9239 (18.99)        13;0  19.0700 (0.81)         22           1
[test_320x240.mp4-1-4]     42.2638 (1.06)     49.8599 (1.14)     47.1788 (1.12)     2.4160 (2.05)     47.4803 (1.11)      3.8516 (4.32)          8;0  21.1960 (0.89)         21           1
[test_320x240.mp4-1-1]     43.0017 (1.08)     54.3128 (1.24)     46.4622 (1.10)     3.0641 (2.61)     45.6213 (1.07)      2.8806 (3.23)          5;1  21.5229 (0.91)         16           1
[test_320x240.mp4-0-0]     43.4649 (1.09)     66.6785 (1.52)     53.9153 (1.28)     7.4268 (6.32)     50.5543 (1.18)     11.7557 (13.19)         7;0  18.5476 (0.78)         17           1
[test_320x240.mp4-2-4]     43.8470 (1.10)     65.7712 (1.50)     55.6573 (1.32)     7.9761 (6.78)     61.6684 (1.44)     14.8573 (16.67)         7;0  17.9671 (0.76)         20           1
[test_320x240.mp4-0-1]     46.7099 (1.17)     53.5070 (1.22)     49.2058 (1.17)     1.5920 (1.35)     48.9321 (1.15)      0.8911 (1.0)           4;4  20.3228 (0.86)         20           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

------------------------------------------------------------- benchmark 'ffmpeg_decode_video_benchmark_test_640x480.mp4': 7 tests -------------------------------------------------------------
Name (time in ms)               Min                 Max                Mean             StdDev              Median                IQR            Outliers     OPS            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
[test_640x480.mp4-2-1]     127.1230 (1.0)      137.4464 (1.0)      130.2180 (1.0)       3.3199 (2.34)     129.3156 (1.0)       3.3550 (1.58)          1;1  7.6794 (1.0)           8           1
[test_640x480.mp4-0-1]     138.5035 (1.09)     164.5596 (1.20)     149.2877 (1.15)     10.9136 (7.70)     146.8460 (1.14)     18.0768 (8.49)          1;0  6.6985 (0.87)          5           1
[test_640x480.mp4-1-4]     149.9544 (1.18)     170.9589 (1.24)     159.6738 (1.23)      9.0782 (6.41)     154.2905 (1.19)     16.0728 (7.55)          2;0  6.2628 (0.82)          7           1
[test_640x480.mp4-2-4]     151.8572 (1.19)     159.0967 (1.16)     155.4907 (1.19)      2.8463 (2.01)     155.7198 (1.20)      5.1922 (2.44)          3;0  6.4313 (0.84)          7           1
[test_640x480.mp4-0-4]     152.4217 (1.20)     156.2592 (1.14)     154.5862 (1.19)      1.4168 (1.0)      154.1098 (1.19)      2.1281 (1.0)           2;0  6.4689 (0.84)          7           1
[test_640x480.mp4-1-1]     154.5291 (1.22)     176.1832 (1.28)     159.7059 (1.23)      7.5032 (5.30)     157.8567 (1.22)      4.1224 (1.94)          1;1  6.2615 (0.82)          7           1
[test_640x480.mp4-0-0]     190.3824 (1.50)     232.0148 (1.69)     217.8044 (1.67)     17.2536 (12.18)    223.0630 (1.72)     24.5422 (11.53)         1;0  4.5913 (0.60)          5           1
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

---------------------------------------------------------- benchmark 'ffmpeg_decode_video_benchmark_test_160x120.mp4': 7 tests -----------------------------------------------------------
Name (time in ms)              Min                Max               Mean            StdDev             Median               IQR            Outliers      OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
[test_160x120.mp4-0-4]     11.8120 (1.0)      23.3611 (1.33)     15.1492 (1.04)     1.8795 (1.54)     14.8747 (1.03)     1.9712 (1.12)         13;2  66.0102 (0.96)         63           1
[test_160x120.mp4-1-4]     12.0402 (1.02)     37.3292 (2.12)     16.6503 (1.14)     3.5744 (2.94)     16.2662 (1.13)     3.7718 (2.15)         11;1  60.0589 (0.87)         70           1
[test_160x120.mp4-2-4]     12.1617 (1.03)     20.8816 (1.19)     15.6181 (1.07)     2.1176 (1.74)     14.8835 (1.03)     2.3251 (1.33)         18;5  64.0284 (0.93)         70           1
[test_160x120.mp4-2-1]     12.3020 (1.04)     17.6063 (1.0)      14.5655 (1.0)      1.2176 (1.0)      14.3828 (1.0)      1.7548 (1.0)          13;0  68.6552 (1.0)          46           1
[test_160x120.mp4-1-1]     12.7449 (1.08)     22.4380 (1.27)     15.2390 (1.05)     1.4606 (1.20)     14.5028 (1.01)     1.9501 (1.11)          7;1  65.6213 (0.96)         63           1
[test_160x120.mp4-0-0]     15.1875 (1.29)     29.7160 (1.69)     20.6005 (1.41)     5.3943 (4.43)     20.2757 (1.41)     9.5309 (5.43)          2;0  48.5425 (0.71)         10           1
[test_160x120.mp4-0-1]     16.1872 (1.37)     36.2697 (2.06)     19.2711 (1.32)     3.6380 (2.99)     18.8951 (1.31)     2.0014 (1.14)          1;1  51.8912 (0.76)         27           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

Also attach the benchmark plots here: [plots.zip](https://github.com/tensorflow/io/files/6519634/plots.zip)

From the results, I do see that `thread_type=2` and `thread_count=1` combination outperforms others in the majority of the test cases. I suspect that ffmpeg implementation of H264 decoder has some issues in utilizing more than 1 thread. I do not yet know what happens in other video codecs, but considering the popularity of H264 format I assume setting `thread_type=2` and `thread_count=1` to be the default option makes sense.

**Comment by yongtang** [comment]
Thanks @kyamagu for the performance tests!

Given the result, can you update the PR to set `thread_type=2` and `thread_count=1` internally and hide the options at the python API level? I think we can leave C++ API part of the options alone.

My reservation with python API is that, once it is added, it is always hard to make a change or remove it in the future  as we want to have some level of API compatibility (C++ API is fine).

We can decide to add options to python API later if we really see they are absolutely needed.

File: tensorflow_io/core/kernels/ffmpeg_kernels.cc [hunk 1]
```diff
@@ -300,7 +302,7 @@ class FFmpegAudioStream : public FFmpegStream {
 
   Status OpenAudio(int64 index) {
     TF_RETURN_IF_ERROR(Open(AVMEDIA_TYPE_AUDIO, index));
-    TF_RETURN_IF_ERROR(OpenCodec());
+    TF_RETURN_IF_ERROR(OpenCodec(FF_THREAD_SLICE, 1));
```
Comment (kvignesh1420) [inline tensorflow_io/core/kernels/ffmpeg_kernels.cc:305]: @kyamagu can you please add a comment mentioning that `FF_THREAD_SLICE` has a defined value of 2? I see that it is coming from `#include "libavcodec/avcodec.h"` but a comment will be helpful for future references.

Comment (kyamagu) [reply tensorflow_io/core/kernels/ffmpeg_kernels.cc:305]: Sure

**Review by kvignesh1420 — APPROVED** [review]
LGTM.

**Review by yongtang — APPROVED** [review]
LGTM

**Pull request merged**